Topics/Tech_Stacks/redux.md

```diff
@@ -0,0 +1,141 @@
+
+However, Redux is not always the best choice for state management. It's best suited for large applications with complex state changes. For smaller applications, you can use React's built-in state management. Complexity is subjective, so it's up to you to decide whether Redux is the right choice for your application.
+
+## How to Install Redux
+
+To install Redux, you need to have Node.js and npm (Node Package Manager) installed on your machine. Once you have them, you can install Redux using npm:
+
+```bash
+npm install redux
+```
+
+For a React application, you will also need to install React-Redux:
+
+```bash
+npm install react-redux
+```
+
+## Core Concepts
+
+### Store
+
+
+In Redux, a store is like a central warehouse that holds all the data your application needs. It's responsible for storing this data, updating it based on specific messages called actions, providing the current data to different parts of your application, and notifying them when any data changes. This centralization helps keep your application organized and ensures that the data flows in a predictable and manageable way.
+
+#### 1. `dispatch`
+
+- The `dispatch` method is used to send actions to the Store, which are processed by reducers to update the state.
+- Example:
+  ```javascript
+  // Dispatching an action
+  store.dispatch({ type: 'INCREMENT' });
+  ```
+
+#### 2. `getState`
+
+- The `getState` method returns the current state held in the Redux Store. It's useful for accessing the state of the application.
+- Example:
+  ```javascript
+  // Accessing the current state
+  const currentState = store.getState();
+  console.log(currentState);
+  ```
+
+#### 3. `subscribe`
+
+- The `subscribe` method allows you to listen for state changes in the Store. It takes a callback function that gets executed whenever the state changes.
+- Example:
+  ```javascript
+  // Subscribing to state changes
+  const unsubscribe = store.subscribe(() => 
+    console.log('State changed:', store.getState())
+  );
+
+  // To unsubscribe from the state changes
+  unsubscribe();
+  ```
+
+
+The Store ensures that state is kept consistent across your application.
+
+### Action
+
+Actions are plain JavaScript objects that represent what happened and what needs to change in the application's state. Each action must have a `type` property, which is a string describing how the state should change.
+
+A typical action may look like this:
+
+```javascript
+{
+  type: 'ADD_TODO',
+  text: 'Complete the Redux tutorial'
+}
+```
+
+Actions are dispatched using the `dispatch` method of the Store.
+
+### Reducer
+
+Reducers are pure functions that take the current state and an action, and return a new state. They describe how the state changes in response to an action. Reducers should be deterministic, meaning given the same input, they should always return the same output.
+
+A reducer for a to-do application might look like this:
+
```

Comment:
Now, I understand that the action is just a javascript object, which represents as the message sending to the store, telling what to update. And The store is like a container that has all the states. However, I don't quite get the concept of reducer. The explanation now might be hard to understand for a person without experience.

Reply:
Generally, in Redux, a reducer is a function that decides how the state of your application changes in response to an action. To put it simply, a reducer acts like a director who takes instructions (actions) and updates the state of the application accordingly. I will add a workflow explaining how to use reducer.